docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,62 @@
+The purpose of this file is to describe the functional and key architecture of
+this project. It is spiced with some rational throughout.
+
+# Slots & Tokens
+
+This library shall provide multiple tokens for the user(s). Any given slot can
+only contain one token at any given time. In case of smartcards the slot is a
+reader whilst the token is the actual smartcard. Since this project won't
+implement a separate tool for "virtual swapping" of tokens, each token is
+associated with its own slot. Thus, the library offers a set of slots with a
+token inside each of these.
+
+In order to allow the creation of new tokens using the pkcs11 means, this
+library presents one more slot/token pair than tokens currently exist. The
+additional token is a blank token that can be initialized using the C_InitToken
+function.
+(Note: This is not implemented yet)
+
+# Objects
+
+Each token can contain an arbitrary amount of keys and other object. (Note:
+other objects are not implemented yet). The keys are TPM-based keys, i.e. they
+are only known to the TPM and also the key operations are executed inside the
+TPM. The other objects are encrypted using hybrid encryption with the outer
+key being sealed to the TPM.
```

Comment:
depends, small ones would be sealed to the TPM big ones use wrapping with a tpm key. I guess that probably doesn't matter and we could always just create key object + wrap.

Reply:
I guess that's true. Though it makes me think if a single code-path would be more beneficial than saving the intermediate key for small payloads...
What do you think ?

Reply:
I don't know how you would get to a single code path unless that code path was just "create key and encrypt blob".

Reply:
Well, instead of going
a) Create intermediate key and seal that and encrypt with that AND
b) Don't create intermediate keys but seal data directly
We could'd just get rid of b) and always create an intermediate key. That's the reduced code-path I'm referring to...
What do you think ?

Reply:
Yeah, we're thinking the same here.